src/cargo/util/errors.rs

```diff
@@ -8,144 +7,100 @@ use std::str;
 use std::string;
 
 use core::TargetKind;
+use util::network::{NetworkError, NetworkErrorKind};
 
 use curl;
 use git2;
-use glob;
 use semver;
 use serde_json;
 use term;
 use toml;
-use url;
 
-pub type CargoResult<T> = Result<T, Box<CargoError>>;
-
-// =============================================================================
-// CargoError trait
-
-pub trait CargoError: Error + Send + 'static {
-    fn is_human(&self) -> bool { false }
-    fn cargo_cause(&self) -> Option<&CargoError>{ None }
-    fn as_error(&self) -> &Error where Self: Sized { self as &Error }
-}
-
-impl Error for Box<CargoError> {
-    fn description(&self) -> &str { (**self).description() }
-    fn cause(&self) -> Option<&Error> { (**self).cause() }
-}
-
-impl CargoError for Box<CargoError> {
-    fn is_human(&self) -> bool { (**self).is_human() }
-    fn cargo_cause(&self) -> Option<&CargoError> { (**self).cargo_cause() }
-}
-
-// =============================================================================
-// Chaining errors
-
-pub trait ChainError<T> {
-    fn chain_error<E, F>(self, callback: F) -> CargoResult<T>
-                         where E: CargoError, F: FnOnce() -> E;
-}
-
-#[derive(Debug)]
-struct ChainedError<E> {
-    error: E,
-    cause: Box<CargoError>,
-}
-
-impl<'a, T, F> ChainError<T> for F where F: FnOnce() -> CargoResult<T> {
-    fn chain_error<E, C>(self, callback: C) -> CargoResult<T>
-                         where E: CargoError, C: FnOnce() -> E {
-        self().chain_error(callback)
+error_chain! {
+    types {
+        CargoError, CargoErrorKind, CargoResultExt, CargoResult;
```

Comment:
I think eventually we'll want to not override the names here, but I like the idea of leaving this here for now to reduce the churn in the PR